File: PainHouse.java
public class PainHouse {
    //Time Complexity-O(n)
    //Space Complexity-O(1)
    public int minCost(int[][] costs) {
        if(costs==null || costs.length==0){
            return 0;
        }
        int n=costs.length;
        //int[][] dp=new int[n][3];

        // dp[n-1][0]=costs[n-1][0];
        // dp[n-1][1]=costs[n-1][1];
        // dp[n-1][2]=costs[n-1][2];

        int colorR=costs[n-1][0];
        int colorB=costs[n-1][1];
        int colorG=costs[n-1][2];

        for(int i=n-2;i>=0;i--){
            int tempR=colorR;
            int tempB=colorB;
            colorR=costs[i][0]+Math.min(colorB,colorG);
            colorB=costs[i][1]+Math.min(tempR,colorG);
            colorG=costs[i][2]+Math.min(tempR,tempB);

            // dp[i][0]=costs[i][0]+Math.min(dp[i+1][1],dp[i+1][2]);
            // dp[i][1]=costs[i][1]+Math.min(dp[i+1][0],dp[i+1][2]);
            // dp[i][2]=costs[i][2]+Math.min(dp[i+1][0],dp[i+1][1]);
        }

        // return Math.min(dp[0][0],Math.min(dp[0][1],dp[0][2]));
        return Math.min(colorR,Math.min(colorB,colorG));
    }
}

File: painHouse.py
# Recursive Paint House

# Time Complexity: O(3 * 2^n)

# Space Complexity: O(3 * 2^n) ~ Tree with 2^n nodes for the 3 choices of colors

# In this approach, we compute the minimum of the tree nodes below the first choices, and finally choose the minimum between the 3 paths

class Solution:
    def minCost(self, costs: List[List[int]]) -> int:
        if costs is None or len(costs) == 0:
            return 0
        caseR = self.helper(costs, 0, 0, 0)
        caseB = self.helper(costs, 0, 1, 0)
        caseG = self.helper(costs, 0, 2, 0)
        
        return min(caseR, min(caseB, caseG))
    
    def helper(self, costs: List[List[int]], r: int, color: int, minCost: int) -> int:
        
        # Base Case
        if r == len(costs):
            return minCost
        # Choosing a Color Case
        
        if (color == 0):
            return min(self.helper(costs, r+1, 1, minCost + costs[r][0]),
                      self.helper(costs, r+1, 2, minCost + costs[r][0]))
        if (color == 1):
            return min(self.helper(costs, r+1, 0, minCost + costs[r][1]),
                      self.helper(costs, r+1, 2, minCost + costs[r][1]))
        if (color == 2):
            return min(self.helper(costs, r+1, 0, minCost + costs[r][2]),
                      self.helper(costs, r+1, 1, minCost + costs[r][2]))
        
        return 99999
